File: backend/schemas/knowledge.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class CreateKnowledgeFolderRequest(BaseModel):
    parent_id: str | None = Field(default=None, alias="parentId")
    name: str = Field(min_length=1, max_length=255)


class UpdateKnowledgeNodeRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    parent_id: str | None = Field(default=None, alias="parentId")

    def has_single_change(self) -> bool:
        return len(self.model_fields_set) == 1 and self.model_fields_set <= {"name", "parent_id"}


class KnowledgeNodeResponse(BaseModel):
    id: str
    parent_id: str | None = Field(serialization_alias="parentId")
    node_type: Literal["folder", "file"] = Field(serialization_alias="nodeType")
    name: str
    status: Literal["uploaded", "processing", "ready", "failed"] | None = None
    mime_type: str | None = Field(default=None, serialization_alias="mimeType")
    byte_size: int | None = Field(default=None, serialization_alias="byteSize")
    created_at: datetime = Field(serialization_alias="createdAt")
    updated_at: datetime = Field(serialization_alias="updatedAt")
    children: list["KnowledgeNodeResponse"] = Field(default_factory=list)


class KnowledgeTreeResponse(BaseModel):
    items: list[KnowledgeNodeResponse]

File: frontend/src/api/knowledge.ts
import { getStoredAccessToken } from '../lib/auth'
import { request, requestForm } from './http'

export type KnowledgeNode = {
  id: string
  parentId: string | null
  nodeType: 'folder' | 'file'
  name: string
  status: 'uploaded' | 'processing' | 'ready' | 'failed' | null
  mimeType: string | null
  byteSize: number | null
  createdAt: string
  updatedAt: string
  children: KnowledgeNode[]
}

type KnowledgeTreeResponse = { items: KnowledgeNode[] }

function authorizationHeader() {
  const accessToken = getStoredAccessToken()
  if (!accessToken) throw new Error('登录已过期，请重新登录')
  return { Authorization: `Bearer ${accessToken}` }
}

export function getKnowledgeTree() {
  return request<KnowledgeTreeResponse>('/api/knowledge/nodes', { headers: authorizationHeader() })
}

export function createKnowledgeFolder(parentId: string | null, name: string) {
  return request<KnowledgeNode>('/api/knowledge/nodes', { body: { parentId, name }, headers: authorizationHeader(), method: 'POST' })
}

export function renameKnowledgeNode(nodeId: string, name: string) {
  return request<KnowledgeNode>(`/api/knowledge/nodes/${encodeURIComponent(nodeId)}`, { body: { name }, headers: authorizationHeader(), method: 'PATCH' })
}

export function moveKnowledgeNode(nodeId: string, parentId: string | null) {
  return request<KnowledgeNode>(`/api/knowledge/nodes/${encodeURIComponent(nodeId)}`, { body: { parentId }, headers: authorizationHeader(), method: 'PATCH' })
}

export function deleteKnowledgeNode(nodeId: string) {
  return request<void>(`/api/knowledge/nodes/${encodeURIComponent(nodeId)}`, { headers: authorizationHeader(), method: 'DELETE' })
}

export function uploadKnowledgeFile(parentId: string | null, file: File) {
  const formData = new FormData()
  if (parentId) formData.set('parentId', parentId)
  formData.set('file', file)
  return requestForm<KnowledgeNode>('/api/knowledge/files', formData, { headers: authorizationHeader(), method: 'POST' })
}
